spring-pulsar/src/test/java/org/springframework/pulsar/observation/ObservationIntegrationTests.java

```diff
@@ -0,0 +1,185 @@
+ * {@link PulsarListenerObservation receive} observations in Spring Pulsar against all
+ * supported Tracing runtimes.
+ *
+ * @author Chris Bono
+ * @see SampleTestRunner
+ */
+public class ObservationIntegrationTests extends SampleTestRunner {
+
+	@SuppressWarnings("unchecked")
+	@Override
+	public SampleTestRunnerConsumer yourCode() {
+		// template -> listener -> template -> listener
+		return (bb, meterRegistry) -> {
+			ObservationRegistry observationRegistry = getObservationRegistry();
+			try (AnnotationConfigApplicationContext applicationContext = new AnnotationConfigApplicationContext()) {
+				applicationContext.registerBean(ObservationRegistry.class, () -> observationRegistry);
+				applicationContext.register(ObservationIntegrationTestAppConfig.class);
+				applicationContext.refresh();
+				applicationContext.getBean(PulsarTemplate.class).send("obs1-topic", "hello");
+				CountDownLatch latch = applicationContext.getBean(ObservationIntegrationTestAppListeners.class).latch;
+				assertThat(latch.await(10, TimeUnit.SECONDS)).isTrue();
+			}
+
+			List<FinishedSpan> finishedSpans = bb.getFinishedSpans();
+			SpansAssert.assertThat(finishedSpans).haveSameTraceId().hasSize(4);
+
+			List<FinishedSpan> producerSpans = finishedSpans.stream()
+					.filter(span -> span.getKind().equals(Kind.PRODUCER)).toList();
+			SpanAssert.assertThat(producerSpans.get(0)).hasTag("spring.pulsar.template.name", "pulsarTemplate");
+			SpanAssert.assertThat(producerSpans.get(1)).hasTag("spring.pulsar.template.name", "pulsarTemplate");
+
+			List<FinishedSpan> consumerSpans = finishedSpans.stream()
+					.filter(span -> span.getKind().equals(Kind.CONSUMER)).toList();
+			SpanAssert.assertThat(consumerSpans.get(0)).hasTagWithKey("spring.pulsar.listener.id");
+			assertThat(consumerSpans.get(0).getTags().get("spring.pulsar.listener.id")).isIn("obs1-id-0", "obs2-id-0");
+			SpanAssert.assertThat(consumerSpans.get(1)).hasTagWithKey("spring.pulsar.listener.id");
+			assertThat(consumerSpans.get(1).getTags().get("spring.pulsar.listener.id")).isIn("obs1-id-0", "obs2-id-0");
+			assertThat(consumerSpans.get(0).getTags().get("spring.pulsar.listener.id"))
+					.isNotEqualTo(consumerSpans.get(1).getTags().get("spring.pulsar.listener.id"));
+
+			MeterRegistryAssert.assertThat(getMeterRegistry())
+					.hasTimerWithNameAndTags("spring.pulsar.template",
+							KeyValues.of("spring.pulsar.template.name", "pulsarTemplate"))
+					.hasTimerWithNameAndTags("spring.pulsar.template",
+							KeyValues.of("spring.pulsar.template.name", "pulsarTemplate"))
+					.hasTimerWithNameAndTags("spring.pulsar.listener",
+							KeyValues.of("spring.pulsar.listener.id", "obs1-id-0"))
+					.hasTimerWithNameAndTags("spring.pulsar.listener",
+							KeyValues.of("spring.pulsar.listener.id", "obs2-id-0"));
+		};
+	}
+
+	@Configuration(proxyBeanMethods = false)
+	@EnablePulsar
+	static class ObservationIntegrationTestAppConfig extends AbstractContainerBaseTests {
```

Comment:
Is it really intentional to extend from that class?

Reply:
Yes as it ensures the PulsarTestContainer is up and running. I want to move that to a utils rather than have to subclass it everywhere.

Reply:
That is even better (interface). I will do this in follow on PR outside of this scope (see #148 )